### PR TITLE
make bigsampler bq output partition configurable

### PR DESCRIPTION
**Pull request by benkonz**
adds a new arg to BigSampler called `bigqueryPartitioning`, defaults to "DAY", which should maintain the same behavior as before. Users can pass in "DAY|HOUR|MONTH|YEAR", as well as NULL if no table partitioning is desired. 

Making this change so that Ratatool works better with Spotify's internal Luigi BigQuery tasks, which use table sharding as partitioning, and when ratatool sets the partitioning to ingestion day, it causes problems with retention.

Tested by outputting [this](https://console.cloud.google.com/bigquery?project=data-quality-spotify&ws=!1m5!1m4!4m3!1sdata-quality-spotify!2sbenk_test_eu!3sbenk_test_eu_20240219) table via this workflow:

```
apiVersion: workflow.data.spotify.com/v1alpha1
kind: Workflow
metadata:
  name: ratatool-internal-examples-stream-days-bigsampler
  namespace: data-quality-spotify
spec:
  resourceID: ratatool-internal-examples.stream.days.BigSampler
  componentID: ratatool-internal
  scheduling:
    schedule: daily
  serviceAccountRef:
    external: contours-test-pipeline@data-quality-spotify.iam.gserviceaccount.com
  docker:
    args:
      - 'wrap-luigi'
      - '--module'
      - 'luigi_tasks'
      - 'BigSampler'
      - '--uri-prefix'
      - 'gs://benk-playground'
      - '--project'
      - 'data-quality-spotify'
      - '--service-account'
      - 'contours-test-pipeline@data-quality-spotify.iam.gserviceaccount.com'
      - '--input-endpoint'
      - 'spotify-people:groups.groups_%Y%m%d'
      - '--output-endpoint'
      - 'data-quality-spotify:benk_test_eu.benk_test_eu_%Y%m%d'
      - '--sample'
      - '0.01'
      - '--partition'
      - '{}'
    terminationLogging: true
    image: 43ea5c916cd5a85623bf0de598da15982c29d8952dbf63a068d10e5b56466e61
  workflowAlertingDisabled: true
```

the `43ea5c916cd5a85623bf0de598da15982c29d8952dbf63a068d10e5b56466e61` docker image is using my local ratatool PR's code via `sbt publishM2`

the linked table has to partitioning and uses the sharding generated by the BigQueryTarget in Luigi

[here](https://console.cloud.google.com/bigquery?project=data-quality-spotify&ws=!1m5!1m4!4m3!1sdata-quality-spotify!2sbenk_test_eu2!3sbenk_test_eu_20240220) is another table that is using the `--bigquery-partitioning` arg to set the partitioning to "MONTH".

### CODE REVIEW

File: ratatool-sampling/README.md [hunk 1]
```diff
@@ -18,15 +18,16 @@ For full details see [BigSample.scala](https://github.com/spotify/ratatool/blob/
...
+  --output=<path>                                     Output file path or BigQuery table
+  [--fields=<field1,field2,...>]                      An optional list of fields to include in hashing for sampling cohort selection
+  [--seed=<seed>]                                     An optional seed used in hashing for sampling cohort selection
+  [--hashAlgorithm=(murmur|farm)]                     An optional arg to select the hashing algorithm for sampling cohort selection. Defaults to FarmHash for BigQuery compatibility
+  [--distribution=(uniform|stratified)]               An optional arg to sample for a stratified or uniform distribution. Must provide `distributionFields`
+  [--distributionFields=<field1,field2,...>]          An optional list of fields to sample for distribution. Must provide `distribution`
+  [--exact]                                           An optional arg for higher precision distribution sampling.
+  [--bigqueryPartitioning=<DAY|HOUR|MONTH|YEAR|null>] An optional arg specifying what partitioning to use for the output BigQuery table, or 'null' for no partitioning. Defaults to DAY.
```
Comment (monzalo14) [inline ratatool-sampling/README.md:30]: Is there a reason to have `null` instead of `NULL` here? I know you're converting case later but the error messages with caps on could confuse the user.

Comment (benkonz) [reply ratatool-sampling/README.md:30]: I can make all the examples lowercase for consistency, since they get normalized to uppercase in the code

File: ratatool-sampling/src/main/scala/com/spotify/ratatool/samplers/BigSampler.scala [hunk 2]
```diff
@@ -101,16 +101,18 @@ object BigSampler extends Command {
...
+        |  [--fields=<field1,field2,...>]                      An optional list of fields to include in hashing for sampling cohort selection
+        |  [--seed=<seed>]                                     An optional seed used in hashing for sampling cohort selection
+        |  [--hashAlgorithm=(murmur|farm)]                     An optional arg to select the hashing algorithm for sampling cohort selection. Defaults to FarmHash for BigQuery compatibility
+        |  [--distribution=(uniform|stratified)]               An optional arg to sample for a stratified or uniform distribution. Must provide `distributionFields`
+        |  [--distributionFields=<field1,field2,...>]          An optional list of fields to sample for distribution. Must provide `distribution`
+        |  [--exact]                                           An optional arg for higher precision distribution sampling.
+        |  [--byteEncoding=(raw|hex|base64)]                   An optional arg for how to encode fields of type bytes: raw bytes, hex encoded string, or base64 encoded string. Default is to hash raw bytes.
```
Comment (monzalo14) [inline ratatool-sampling/src/main/scala/com/spotify/ratatool/samplers/BigSampler.scala:113]: Nit: Is it possible to separate formatting changes into a different PR?

**Review by monzalo14 — APPROVED** [review]
I'd love if you could separate formatting changes from actual changes, but LGTM!

File: ratatool-sampling/src/main/scala/com/spotify/ratatool/samplers/BigSampler.scala (hunk 2 above)
Comment (benkonz) [reply ratatool-sampling/src/main/scala/com/spotify/ratatool/samplers/BigSampler.scala:113]: not really, since I'm aligning the new whitespace with the field being added in this PR
